smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/waiters/JMESPathVisitor.kt

```diff
@@ -0,0 +1,412 @@
+    // Maps the expression result in leftName into a new collection using the right expression.
+    private fun mappingBlock(right: JmespathExpression, left: Variable): Variable {
+        when (right) {
+            is CurrentExpression -> return left // Nothing to map
+        }
+        when (left.shape) {
+            is CollectionShape -> {
+                val outerName = uniqueTempVarName("projection")
+                val original = Variable("original", false, model.expectShape(left.shape.member.target))
+                val bufferWriter = BufferWriter(writer)
+                var transformedVar: Variable? = null
+                bufferWriter.record { writer ->
+                    transformedVar = childBlock(right, original)
+                }
+                val optionalityMark = "?".takeIf { left.isOptional } ?: ""
+                writer.openBlock(
+                    "let \$L: [\$L]\$L = \$L\$L.compactMap { original in", "}",
+                    outerName,
+                    transformedVar!!.baseSwiftSymbol(symbolProvider),
+                    optionalityMark,
+                    left.name,
+                    optionalityMark
+                ) {
+                    bufferWriter.playback()
+                    writer.write("return \$L", transformedVar!!.name)
+                }
+                val returnMember = MemberShape.builder()
+                    .id("smithy.swift.synthetic#mappedCollection\$member")
+                    .target(transformedVar!!.shape)
+                    .build()
+                val returnType = ListShape.builder()
+                    .id("smithy.swift.synthetic#mappedCollection")
+                    .member(returnMember)
+                    .build()
+                return Variable(outerName, left.isOptional, returnType)
+            }
+            else -> throw Exception("Mapping a non-collection shape: ${left.shape}")
+        }
+    }
+
+    // Accesses a field on the parent variable & returns a variable containing the
+    // resulting expression.
+    private fun subfield(expression: FieldExpression, parentVar: Variable): Variable {
+        when (parentVar.shape) {
+            is StructureShape -> {
+                val parentMembers = parentVar.shape.members()
+                val subfieldMember = parentMembers.first { it.memberName == expression.name }
+                val subfieldShape = model.expectShape(subfieldMember.target)
+                val subfieldName = symbolProvider.toMemberName(subfieldMember)
+                val fieldOperator = "?.".takeIf { parentVar.isOptional } ?: "."
+                //
+                // Because all fields on Swift models are currently optional, every subfield value will also be
+                // optional.  Later we may have to replace this with logic to determine actual optionality,
+                // so:
+                // val subfieldIsOptional = !subfieldMember.hasTrait<RequiredTrait>()
+                // is temporarily replaced with:
```

Comment:
great context!

Reply:
I'm sure there will be bugs, but flipping this switch should be all that's needed to support non-optional fields on model objects somewhere down the line